File: 459.重复的子字符串.cpp
/*
 * @lc app=leetcode.cn id=459 lang=cpp
 *
 * [459] 重复的子字符串
 */

// @lc code=start
class Solution {
public:
    bool repeatedSubstringPattern(string s) {
        int next[s.size()];
        getNext(&next[0], s);
        int length = next[s.size() - 1];
        return (s.size() % (s.size() - length) == 0) && (length != 0);
    }
    void getNext(int *next, string s) {
        int j = 0; next[0] = 0;
        for (size_t i = 1; i < s.size(); i++) {
            while(j > 0 && s[j] != s[i]) {
                j = next[j - 1];
            }
            if(s[j] == s[i]) {
                j++;
            }
            next[i] = j;
        }
    }

};
// @lc code=end

File: 459.重复的子字符串.java
/*
 * @lc app=leetcode.cn id=459 lang=java
 *
 * [459] 重复的子字符串
 */

// @lc code=start
class Solution {
    public boolean repeatedSubstringPattern(String s) {
        return true;
    }

    private int[] getNextArr(String s) {
        char[] t = s.toCharArray();
        int[] next = new int[t.length];
        // 初始化两指针
        int j = -1;
        next[0] = j;

        for (int i = 1; i < arr.length; i++) {
            if(t[i] == t[j + 1]) {
                j++;
            }
            next[i] = j;
            while(j >= 0 && t[i] != t[j + 1]) {
                j = next[j];
            }
        }
    }
}
// @lc code=end

